File: web/src/app/potassium-argon-age-calculation/services/Experiment.ts
export interface Cycle {
    cycle: string;
    mass: number;
    peak: string;
    measure: string;
    intensity: number;
    time: number;
}

export interface Experiment {
    cycles: Cycle[];
    analysis_date: string;
    spectrum: string;
    sample_id: string;
    type: string;
    spectrum_user_name: string;
    file_name: string;
}

export interface PotassiumArgonAgeParameter {
    delta: number;
    gramsOfK: number;
    weight: number;
    ar40_divides_ar38_t: number;
    ar36_divides_ar38_t: number;
    experiments: { rawFile: File } | Experiment [];
}

File: apps/domain/experiment.py
from dataclasses import dataclass, replace

from domain.cycle import Cycle
from statistics import mean


@dataclass
class Experiment:
    cycles: list[Cycle]
    analysis_date: str
    spectrum: str
    sample_id: str
    type: str
    spectrum_user_name: str
    file_name: str

    def mean(self, removing_blank_mean=0):
        intensities = [cycle.intensity - removing_blank_mean for cycle in self.cycles]
        return mean(intensities)

    def filter_corrected_cycles(self):
        experiment = replace(self)
        experiment.cycles = [cycle for cycle in self.cycles if cycle.is_corrected()]
        return experiment

    def filter_byAr36(self):
        experiment = replace(self)
        experiment.cycles = [cycle for cycle in self.cycles if cycle.is_Ar36()]
        return experiment

    def filter_byAr38(self):
        experiment = replace(self)
        experiment.cycles = [cycle for cycle in self.cycles if cycle.is_Ar38()]
        return experiment

    def filter_byAr40(self):
        experiment = replace(self)
        experiment.cycles = [cycle for cycle in self.cycles if cycle.is_Ar40()]
        return experiment
